### ISSUE TITLE
Analyzer reports: "Invalid override. The type of 'x' isn't a subtype of 'x'"

### ISSUE THREAD

**Issue by DanTup** [dart-lang/sdk#28895]
Copying this from https://github.com/Dart-Code/Dart-Code/issues/261 as I don't think it's specific to Dart Code (however, I can't explain why Atom or DartAnalyzer wouldn't show the same issue).

[Here is a smaller repro than attached to the Dart Code issue](https://github.com/dart-lang/sdk/files/801144/repro.zip), it's basically:

## lib/a.dart
```dart
import 'package:repro/b.dart';

class Renderer
{

}

class Mesh extends Renderable
{
    @override
    void render(Renderer r)
    {

    }
}
```

## lib/b.dart
```dart
import 'package:repro/a.dart';

abstract class Renderable
{
    void render(Renderer r);
}
```

There's also a pubspec with just a name and I've run `pub get`.

The error reported is:

```text
file: 'file:///c%3A/Users/danny/Desktop/repro/repro/lib/a.dart'
severity: 'Error'
message: 'Invalid override. The type of 'Mesh.render' ('(Renderer) → void') isn't a subtype of 'Renderable.render' ('(Renderer) → void').'
at: '11,5'
source: 'dart'
```

There's only one `Renderer` class so it seems like these should be the same and the error is incorrect, however the files reference each other, I'm not sure if that's supported?

The log from the AS server is this:

```text
[11:47:34]: ==> {"id":"1","method":"server.setSubscriptions","params":{"subscriptions":["STATUS"]}}
[11:47:34]: ==> {"id":"2","method":"analysis.setAnalysisRoots","params":{"included":["c:\\Users\\danny\\Desktop\\repro\\repro"],"excluded":[]}}
[11:47:34]: <== {"event":"server.connected","params":{"version":"1.17.0","pid":12036,"sessionId":""}}
[11:47:34]: <== {"id":"1"}
[11:47:34]: <== {"event":"server.status","params":{"pub":{"isListingPackageDirs":true}}}
[11:47:34]: <== {"event":"server.status","params":{"pub":{"isListingPackageDirs":false}}}
[11:47:35]: <== {"id":"2"}
[11:47:35]: <== {"event":"server.status","params":{"analysis":{"isAnalyzing":true}}}
[11:47:35]: <== {"event":"analysis.errors","params":{"file":"c:\\Users\\danny\\Desktop\\repro\\repro\\lib\\a.dart","errors":[]}}
[11:47:35]: <== {"event":"analysis.errors","params":{"file":"c:\\Users\\danny\\Desktop\\repro\\repro\\lib\\a.dart","errors":[{"severity":"ERROR","type":"COMPILE_TIME_ERROR","location":{"file":"c:\\Users\\danny\\Desktop\\repro\\repro\\lib\\a.dart","offset":113,"length":39,"startLine":11,"startColumn":5},"message":"Invalid override. The type of 'Mesh.render' ('(Renderer) → void') isn't a subtype of 'Renderable.render' ('(Renderer) → void').","code":"strong_mode_invalid_method_override","hasFix":false}]}}
[11:47:35]: <== {"event":"analysis.errors","params":{"file":"c:\\Users\\danny\\Desktop\\repro\\repro\\lib\\a.dart","errors":[{"severity":"ERROR","type":"COMPILE_TIME_ERROR","location":{"file":"c:\\Users\\danny\\Desktop\\repro\\repro\\lib\\a.dart","offset":113,"length":39,"startLine":11,"startColumn":5},"message":"Invalid override. The type of 'Mesh.render' ('(Renderer) → void') isn't a subtype of 'Renderable.render' ('(Renderer) → void').","code":"strong_mode_invalid_method_override","hasFix":false}]}}
[11:47:35]: <== {"event":"analysis.errors","params":{"file":"c:\\Users\\danny\\Desktop\\repro\\repro\\lib\\b.dart","errors":[]}}
[11:47:35]: <== {"event":"analysis.errors","params":{"file":"c:\\Users\\danny\\Desktop\\repro\\repro\\lib\\b.dart","errors":[]}}
[11:47:35]: <== {"event":"analysis.errors","params":{"file":"c:\\Users\\danny\\Desktop\\repro\\repro\\lib\\b.dart","errors":[]}}
[11:47:35]: <== {"event":"server.status","params":{"analysis":{"isAnalyzing":false}}}
```

**Comment by lrhn** [comment]
How are you referring to `a.dart` when you run or analyze it?
If you refer to it as `file:///..../a.dart`, then it will be considered a different library than `package:repro/a.dart`. That means that the `Renderer` in the entry-point library is different from the one imported by `b.dart`.
So either run/analyze using a package URI or stop doing absolute imports. I recommend the latter in any case, it's an unnecessary complication to hard-code your own package name into internal references when it's not necessary. If nothing else, it makes it harder to rename the package.

That is, either run/analyze it as `dart package:repro/a.dart` or try changing the imports from being absolute package-URIs to being relative: `import "a.dart";` and `import "b.dart";`. If the problem persists, then it's something else.

**Comment by DanTup** [comment]
@lrhn 

> The problem seems to be how you refer to your entry-point and the relation between the libraries.
> 
> If you refer to a.dart it as file:///..../a.dart (which you do, it says file: 'file:///c%3A/Users/danny/Desktop/repro/repro/lib/a.dart'), 

That filename is what comes back from the analysis server and then gets passed to VS Code in the errors list; nowhere in the repro is importing it like this (nor do we pass it to the analyzer; only the workspace root). That path is (presumably) constructed by the analysis service based on the analysis root we pass to it; I don't believe there's anything I can change here to fix this.

> As for Dart-Code, they might want to refer to files that are part of a package by its package-URI, not its file-location, if that is possible.

Dart Code doesn't invoke the analyzer directly; it spawns the analysis server from the snapshot in the SDK and then passes it the root folder. We don't parse any of the code at all. We just say "here's a folder of Dart code" and the analysis server gives us all the errors; how pubspec/lib/imports are handled are all internal to the analysis service. I believe this issue could be repro'd entirely without Dart Code (though I'm surprised it apparently doesn't occur in Atom, which uses the same service; though I don't have Atom to test it myself).

It's possible this code is incorrect and should be using relative paths; though the author expected it to work (and at first glance, I would to; though I'm not all that experience with Dart). I'm struggling to see how I can fix this in Dart Code though :(

**Comment by lrhn** [comment]
That sounds like it's probably an analyzer problem then.

If the analyzer sees the `pubspec.yaml` with `name: repro` and a `lib/a.dart` file, it should probably treat `a.dart` as having URI `package:repro/a.dart` when analyzing it. If it doesn't, problems like this will occur.
I'm tagging this as an analyzer issue.

The error report does need to show the file name - the analysis server can handle large directories with multiple package structures, so a package URI doesn't uniquely identify the file with the problem.

**Comment by bwilkerson** [comment]
> If the analyzer sees the `pubspec.yaml` with name: `repro` and a `lib/a.dart` file, it should probably treat `a.dart` as having URI `package:repro/a.dart` when analyzing it. If it doesn't, problems like this will occur.

Yes, that's exactly what analysis server does. And as a consequence, we advise people to **always** use `package:` URIs when referring to libraries inside the `lib` directory of a package.

**Comment by bwilkerson** [comment]
I can't reproduce this behavior on bleeding edge. I created a directory structure containing the following:
```
repro
  lib
    a.dart
    b.dart
  pubspec.yaml
```
where `a.dart` and `b.dart` are identical to what's above, and `pubspec.yaml` contains
```
name: repro
version: 0.0.1
description: A test package.
```
I then ran `pub get` to produce a `.packages` file and a `pubspec.lock` file.

When I run the command-line analyzer on `a.dart` I get "No issues found!", and when I load `repro` in IntelliJ I see no errors being reported.

Could you enable the instrumentation log file, load a project like this into Dart Code, and then send me the log file? (Private e-mail is fine if you don't want to attach it.) That will let me see whether there's a mismatch between client and server.

**Comment by DanTup** [comment]
> Yes, that's exactly what analysis server does.

Yeah, I figured this most generally work or there would be loads of issues!

> Could you enable the instrumentation log file, load a project like this into Dart Code, and then send me the log file? 

Yep, attached is the log for the zip file I attached at the top.

[analyzer.instrum.txt](https://github.com/dart-lang/sdk/files/801399/analyzer.instrum.txt)

I'm puzzled about why it only seems to happen in Dart Code; it doesn't seem like Dart Code is doing much that could influence this, but the original reporter also tested in Atom and it didn't happen!

**Comment by bwilkerson** [comment]
Yeah, I'm puzzled too.

Unfortunately, the instrumentation log doesn't show anything that gives me any clues. There are only two requests, both perfectly reasonable, and there's nothing else that Dart Code needed to do. It might be interesting to get a similar log from Atom for comparison, but I honestly don't expect to see anything different.

Just for completeness...

Do you and the user both have a `.packages` file in `repro`?

Do either you, or the user, have any files or directories in either `repro` or any parent directory with any of the following names: `.analysis_options`, `analysis_options.yaml`, `WORKSPACE`, `READONLY`, or `.jiri_root`? If so, which?

**Comment by DanTup** [comment]
> Do you and the user both have a .packages file in repro?

Yep; it was in the zip file up top, but here's the contents:

```text
# Generated by pub on 2017-02-25 11:43:46.036993.
repro:lib/
```

> Do either you, or the user, have any files or directories in either repro or any parent directory with any of the following names: .analysis_options, analysis_options.yaml, WORKSPACE, READONLY, or .jiri_root? If so, which?

I don't believe so, but just to be sure I moved the `repro` folder to the root of a drive (`m:\repro`) and tried again and see the same behaviour upon opening:

```text
[19:48:51]: ==> {"id":"1","method":"server.setSubscriptions","params":{"subscriptions":["STATUS"]}}
[19:48:51]: ==> {"id":"2","method":"analysis.setAnalysisRoots","params":{"included":["m:\\repro"],"excluded":[]}}
[19:48:51]: <== {"event":"server.connected","params":{"version":"1.17.0","pid":3368,"sessionId":""}}
[19:48:51]: <== {"id":"1"}
[19:48:51]: <== {"event":"server.status","params":{"pub":{"isListingPackageDirs":true}}}
[19:48:51]: <== {"event":"server.status","params":{"pub":{"isListingPackageDirs":false}}}
[19:48:52]: <== {"id":"2"}
[19:48:52]: <== {"event":"server.status","params":{"analysis":{"isAnalyzing":true}}}
[19:48:52]: <== {"event":"analysis.errors","params":{"file":"m:\\repro\\lib\\a.dart","errors":[]}}
[19:48:52]: <== {"event":"analysis.errors","params":{"file":"m:\\repro\\lib\\a.dart","errors":[{"severity":"ERROR","type":"COMPILE_TIME_ERROR","location":{"file":"m:\\repro\\lib\\a.dart","offset":113,"length":39,"startLine":11,"startColumn":5},"message":"Invalid override. The type of 'Mesh.render' ('(Renderer) → void') isn't a subtype of 'Renderable.render' ('(Renderer) → void').","code":"strong_mode_invalid_method_override","hasFix":false}]}}
[19:48:52]: <== {"event":"analysis.errors","params":{"file":"m:\\repro\\lib\\a.dart","errors":[{"severity":"ERROR","type":"COMPILE_TIME_ERROR","location":{"file":"m:\\repro\\lib\\a.dart","offset":113,"length":39,"startLine":11,"startColumn":5},"message":"Invalid override. The type of 'Mesh.render' ('(Renderer) → void') isn't a subtype of 'Renderable.render' ('(Renderer) → void').","code":"strong_mode_invalid_method_override","hasFix":false}]}}
[19:48:52]: <== {"event":"analysis.errors","params":{"file":"m:\\repro\\lib\\b.dart","errors":[]}}
[19:48:52]: <== {"event":"analysis.errors","params":{"file":"m:\\repro\\lib\\b.dart","errors":[]}}
[19:48:52]: <== {"event":"analysis.errors","params":{"file":"m:\\repro\\lib\\b.dart","errors":[]}}
[19:48:52]: <== {"event":"server.status","params":{"analysis":{"isAnalyzing":false}}}
```

> I can't reproduce this behavior on bleeding edge

Oh crap, I just realised when looking at the files that I totally neglected to mention that strong-mode is enabled via `.analysis_options`:

```yaml
analyzer:
  strong-mode: true
```

It was in the zip, but I didn't mention it in the text so if you tried building your own repro you probably missed this.

I'm not sure how to enable logging in Atom; but if you think the log would be useful to compare, I'm happy to install it and try and figure it out. Since I haven't repro'd it myself I can't say for sure the OP didn't do something differently and maybe it can be repro'd there; so the "only happening in Dart Code" thing might be a red herring.

**Comment by lrhn** [comment]
> Yes, that's exactly what analysis server does. And as a consequence, we advise people to always use package: URIs when referring to libraries inside the lib directory of a package.

I don't see why using `package:` URIs should be an advantage when a relative URI is just as correct, and shorter. I generally recommend that you don't use `package:` URIs for references inside the same lib directory.
What advantage is it that you get?

**Comment by bwilkerson** [comment]
I enabled strong-mode, but as I expected there was no difference.

**Comment by DanTup** [comment]
Could it be OS-related? I don't have easy access to a non-Windows machine to test, though maybe you have access to a Windows one?

Using the zip file I attached is probably easiest in case I described anything else incorrectly; just opening that folder in Dart Code immediately triggers the error.

If there's anything else you can think of that I can do to help debug, let me know!

**Comment by DanTup** [comment]
I opened the same repro folder on both my PC and Mac, both using 1.24.2 SDK. The PC reports this error and the Mac does not.

So I wonder whether there's some code that considers both the package import and the relative path as the same file that isn't working correctly on Windows (maybe because of path separators?).

**Comment by bwilkerson** [comment]
If it is a Windows specific issue that would explain why I couldn't reproduce it on my Mac.

> So I wonder whether there's some code that considers both the package import and the relative path as the same file that isn't working correctly on Windows (maybe because of path separators?).

That sounds sort of plausible, but I'm not sure where that would happen. Analyzer resolves URIs against URIs and then uses a platform-aware algorithm to convert the resulting URI to a file path. If there isn't something interesting about this case, I would have expected a lot more reports of problems from Windows users.

**Comment by DanTup** [comment]
I thought that too. Do any of you have access to debug this on a Windows machine? I don't have the code running on mine currently, but it probably wouldn't be hard to sort out if someone could give me some pointers on how to debug, I could see if I can get anywhere.

**Comment by DanTup** [comment]
This appears to be fixed by #32133 (normalisation of drive letters on windows, the fix for #32095).

**Comment by bwilkerson** [comment]
> Do any of you have access to debug this on a Windows machine?

Unfortunately, no.

**Comment by DanTup** [comment]
@bwilkerson That was a pretty old comment you replied to :) I believe this issue is fixed along with the other similar issues I sent a PR for :-)

**Comment by bwilkerson** [comment]
Yes :-) but the info about our lack of access to a Windows machine isn't likely to change soon.

**Comment by DanTup** [comment]
I don't think it's a big issue for this case, if the PR I sent is good then I think this is resolved. Are there CI builds/test runs on Windows to ensure I didn't totally smeg everything?

I don't mind digging into Windows-specific issues where I can; I tend to be quite split across Windows and Mac lately anyway.

**Comment by bwilkerson** [comment]
Yes, we have bots that run everything on all three platforms: Linux, Mac, and Windows. But the windows bots are not represented in the subset of try bots, so you won't know you've broken windows until the build fails.

**Comment by DanTup** [comment]
Good to know, ta!

**Comment by DanTup** [comment]
Re-opening this since the fix mentioned above was reverted and this might be a much easier case to address/fix than #32095 so it'd make sense to get out of the way first.

**Comment by DanTup** [comment]
@devoncarew @bwilkerson This seems to be cropping up much more recently. I'm happy to help out if I can - previously I tried to create some Windows tests that failed for the same reasons, but got stuck because the memory file system used in tests is effectively case-sensitive. If anyone has ideas about the best way to fix that, I don't mind having a go.

(though I'm still slightly blocked by https://github.com/dart-lang/sdk/issues/32702... have managed to get down to the command that's failing, but not figured out why!)

**Comment by escamoteur** [comment]
I'm too are now affected. You can try it with https://github.com/escamoteur/making_flutter_reactive-/tree/problem_with_analyzer

when running the App or running flutter analyze everything is ok but in VS code I shows two errors

In homepage.dart
![image](https://user-images.githubusercontent.com/16635729/38936241-a7bfd8e0-4320-11e8-85e0-4fb714be2d08.png)

![image](https://user-images.githubusercontent.com/16635729/38936254-b58d791e-4320-11e8-9f37-b05fde0cfb34.png)

**Comment by devoncarew** [comment]
@DanTup, it sounds like the fix was correct was correct, but we saw some test failures associated with it. Can you in-line the test failures, or reference them here? I think we should look at addressing them and re-applying the fix.

**Comment by DanTup** [comment]
@devoncarew We don't have tests to cover this; when we made the original fix we shipped it because everything passed (we subsequently discovered that some Windows tests were being skipped, but even when running them, none of the failures relate to this).

I'd like to make some failing tests for this (and for the issues introduced by the "fix" we rolled back), but both of them rely on being able to write tests where the file-system is case-insensitive (which isn't the case for the unit tests) or integration tests where we can ensure the VM and whatever's driving the analysis server are using different cased drive letters.

I'll add to the list to discuss tomorrow and maybe we can come up with a plan.

@escamoteur Mixing relative and `package:` imports definitely makes things worse, but I don't think all of the issues are caused by that so you may still hit some issues.

**Comment by bwilkerson** [comment]
I don't think there would be a problem with making `MemoryResourceProvider` (and it's associated classes) consistent with the platform (such as being case sensitive when the underlying platform is case sensitive). That seems to me like the easiest path forward.

**Comment by DanTup** [comment]
> That seems to me like the easiest path forward.

I thought that'd be an easy change to make, but I think it was full of `Map`s to look things up, so it may mean replacing all of those with something we can have a case-insensitive key for?

Also, to confuse things, new Macs seem to also be case-insensitive (though don't hit this issue generally because there isn't a random part of the path being exposed with differing casing in different places) so I'm not sure how complicated this will get (for ex. if I have both a case-sensitive and case-insensitive filesystem on my Mac and run Dart programs from both, should they honour the sensitivity for each drive? Could be a can of worms!)

**Comment by DanTup** [comment]
> Hm. I'm running 10.13.3, and at least in a console window that appears to be true. All of our tests are passing, though. Raises the interesting question of whether case insensitivity is really the issue here.

I think it's just luck that on Mac all the paths that come from the system are consistent, but on Windows the drive letters are not. Does make me wonder whether we can repro the issue by faking the casing of the path in setAnalysisRoots to something different to what the underlying folders were cased though. I'll have a quick play tomorrow and see where I get.

**Comment by DanTup** [comment]
Ok, I repro'd on a Mac really easily (not sure why I hadn't tried this sooner!).. Took the zip file from the very top of this post, and I made this in Dart Code where we send analysis roots:

```
newRoots[0] = newRoots[0].toUpperCase();
```

This results in the path being used by the editor to not match the one that the analysis server is using for `package` URIs, and we get the original error:

```js
{
	"resource": "/USERS/DANTUP/DESKTOP/REPRO/lib/a.dart",
	"owner": "dart",
	"code": "strong_mode_invalid_method_override",
	"severity": 8,
	"message": "Invalid override. The type of 'Mesh.render' ('(Renderer) → void') isn't a subtype of 'Renderable.render' ('(Renderer) → void').",
	"source": "dart",
	"startLineNumber": 11,
	"startColumn": 5,
	"endLineNumber": 11,
	"endColumn": 44
}
```

So, if we have integration tests that use the real file system (and not the in-memory one), a failing test should be pretty easy. If not, we might still need to change the in-memory file system to be case-insensitive to do that.

Fixing things is its own can of worms. What casing of paths do we send back to the client - the case they sent to server or the case we think it is? And how do we handle mixed-case-sensitivity (you can import a Dart file from another drive)?

We could try a partial fix similar to what we did last time, which is to just normalise Windows drive letters in one direction *everywhere* and inform clients they have to deal with that - that all drive letters out of the server will be upper(/lower)case regardless of what they sent (or even, that they must always send uppercase).

I'm now wondering if me normalising drive letters to uppercase in Dart Code would actually fix this (well, at least for the Windows users hitting it casually) - VS Code currently normalises to lowercase (for the editor, sadly it's uppercase for the debugger!) if the SDK is consistently getting uppercase drive letters (it's hard to say without knowing where it gets that from). If this turns out to be a huge job, I can give it a go - I *think* that in all cases where VS Code is normalising my drive letter I'll be accessing the `fsPath` property, so it may be an easy job to map them all (and possibly even have a validation script or something to check for this).

**Comment by DanTup** [comment]
> I'm now wondering if me normalising drive letters to uppercase in Dart Code would actually fix this (well, at least for the Windows users hitting it casually)

![paths](https://user-images.githubusercontent.com/1078012/38982263-4d7144ac-43b9-11e8-8cd0-1e63c3c758c3.png)

I don't know how risky this is (for ex. if there are any people currently getting lowercase drive letters in the analysis server, this will break things for them) but I've put the changes on a branch behind a flag and given a custom build to @escamoteur. If all goes well, this may be a quicker way to solve the users issues without having to divert a load of effort here.

**Comment by KiddoV** [comment]
Got the same issue on Windows. Any update or work-around on this?

**Comment by DanTup** [comment]
@KiddoV can you confirm where you're seeing this?

**Comment by KiddoV** [comment]
@DanTup I ran into this issue today on one of my files right after running `dart run build_runner build`. It has been working fine up until now.

Look at the difference in path casing in the diagnostic message:
```txt
The argument type 'AppQueryParams (where AppQueryParams is defined in E:\github-projects\...)' 
can't be assigned to the parameter type 'AppQueryParams (where AppQueryParams is defined in E:\Github-Projects\...)'. 
dart[argument_type_not_assignable]
```

**Comment by DanTup** [comment]
@KiddoV out of interest, which shell are you running it in, and is the casing of your working directory different to what you see in explorer?

For example if `pwd` shows `github-projects`, does running `cd ../Github-Projects` and then running again prevent the error (I'm trying to understand where the mismatch comes from).

@bwilkerson @scheglov @davidmorgan interested in your thoughts here. I tried to fix this in server a few times in the past but it turned out to be very complicated (there are many places in server that treat string paths as case-sensitive keys and making them all aware of the actual case sensitivity of each volume/folder would be very complicated). The usual issue that comes up is the drive letter casing being different and that's fairly easy to just always normalize in one direction (we tend to use uppercase).

However the path is trickier. In Dart-Code, we detect if the opened workspace folders have different casing to their native case and show a warning stating that it may cause issues and you should re-open the folder using the File -> Open dialog.

My feeling is that fixing it in the analyzer will be a huge job. But having to leak normalization (or detection of "unnatural casing") into each tool that uses the analyzer is not ideal.

(there was also some recent discussion around this in the context of IntelliJ in https://github.com/dart-lang/sdk/issues/63819)

**Comment by davidmorgan** [comment]
Yeeeah this sounds like a tough one.

Just the part I'm familiar with: `package:watcher` does not correctly fold case on Windows/Mac.

https://github.com/dart-lang/tools/issues/2271

So case-changing renames are one place the analysis server can't currently get things right.

Happy to push on that issue if it's useful, if so please comment on it :)

**Comment by KiddoV** [comment]
@DanTup 

I'm using `VS Code` with the integrated `PowerShell` terminal.

1. Terminal `pwd` output:
```txt
Path                                    
----                                    
E:\Github-Projects\...
```
2. Physical Explorer path casing: `E:\Github-Projects\...` (matches pwd).
3. Casing test: I tried explicitly running cd `e:\github-projects\...` (lowercase) and re-running `dart run build_runner build`, but the analyzer error persisted.
4. Analysis Server restart: Restarting the `Dart Analysis Server (Dart: Restart Analysis Server)` while in this state still didn't resolve the diagnostic error for this file.

Let me know if anything else I can do for testing...

**Comment by DanTup** [comment]
@KiddoV that's interesting - I wonder where the lowercase version is coming from then. @davidmorgan I do see some calls to `path.canonicalize` in the build_runner repo, but I don't know if they make it to the server like that.

While fixing the wactcher package would be nice, I'm not sure on its own it'll fix a lot (FWIW VS Code has the same issue, it doesn't tell LSP servers about case-only renames, and continues to use the old casing for all future requests in the session, which also causes issues when trying to fix lints about file casing 😔).

**Comment by bwilkerson** [comment]
If we don't know what's causing the problem in this most recent report, that's probably what we need to start with. An instrumentation log will presumably preserve case, so it might tell us when we're seeing a case-only difference in paths.

**Comment by scheglov** [comment]
Ideally, we should not do anything in the analyzer.
The analyzer should use the exact paths that the OS reports for listing directories.

Inputs from outside could be in different cases, because on Windows this is allowed.
But these inputs should be canonicalized before entering the analyzer.

We have some checks against this, e.g. `_ensureAbsoluteAndNormalized` in resource providers.

**Comment by DanTup** [comment]
@bwilkerson 

> An instrumentation log will presumably preserve case, so it might tell us when we're seeing a case-only difference in paths.

If I understand correctly, the issue here is when running `build_runner`. I presume it's not using server and is using analyzer, but I'm not that familiar with it. @davidmorgan is there a log or anything that might help understand the casing of paths being passed around/into analyzer here?


@scheglov 

> Inputs from outside could be in different cases, because on Windows this is allowed.

Not just on Windows, current macOS tends to be case-insensitive too. But you according to [this comment](https://github.com/dart-lang/sdk/issues/63819#issuecomment-4980774886) it can vary per-directory 🙀


> But these inputs should be canonicalized before entering the analyzer.
> 
> We have some checks against this, e.g. `_ensureAbsoluteAndNormalized` in resource providers.


What do you mean by "canonicalized"? `path.canonicalize()` just lowercases on Windows (which might be what's happening here). So possibly you mean translated to the "natural" case on disk. But doing that on every path could be very expensive. That's why in Dart-Code we only check the workspace folders at startup and warn. We then assume any path below that we'll get naturally in its natural case.


I think it's unfortunate to have to duplicate this effort in every consumer of the analyzer package, but I also suspect it's the probably the simplest/most feasible solution.

**Comment by scheglov** [comment]
> > But these inputs should be canonicalized before entering the analyzer.
> > We have some checks against this, e.g. `_ensureAbsoluteAndNormalized` in resource providers.
> 
> What do you mean by "canonicalized"? `path.canonicalize()` just lowercases on Windows (which might be what's happening here). So possibly you mean translated to the "natural" case on disk.

Yes, "on disk" is what meant.
I used the word "canonical" is a wide sense.

> But doing that on every path could be very expensive. That's why in Dart-Code we only check the workspace folders at startup and warn. We then assume any path below that we'll get naturally in its natural case.

I thought that if we "canonicalize" inputs to "on disk" versions, and then as we read directories we already get on-disk path, then it is not super expensive. We usually get only a handful of input paths for DAS as roots.

> I think it's unfortunate to have to duplicate this effort in every consumer of the analyzer package, but I also suspect it's the probably the simplest/most feasible solution.

Well, working with paths is not the analyzer's area.
This is why we insist that we are given absolute normalized paths.

**Comment by davidmorgan** [comment]
build_runner doesn't have any direct link with the analyzer that's running in the IDE; it uses its own private one with no shared state.

So from the analyzer's point of view it only writes files.

**Comment by DanTup** [comment]
Oh, I might have misunderstood where @KiddoV was seeing these errors. I assumed in the output from running build_runner, but now I'm not so sure. @KiddoV are you seeing these in the output of the build_runner command, or is running `build_runner` causing the diagnostics in the IDE?

Can you reproduce the issue on a project that can be shared? (I did try to repro with the build_runner example project with no luck, but it's a very trivial example that doesn't generate much code).

**Comment by davidmorgan** [comment]
It looks like there may be a build_runner issue after all.

It uses

https://pub.dev/documentation/path/latest/path/canonicalize.html

which converts paths to lowercase on Windows.

So when it writes files, it writes them with an all-lowercase path.

If I remember correctly how package:watcher behaves, this will lead to watch events firing with an all-lowercase path.

**Comment by DanTup** [comment]
Yeah, that's what I wondered [above](https://github.com/dart-lang/sdk/issues/28895#issuecomment-5606245567). In which case, maybe a fix in `watcher` to generate events with the natural case of the path would also help (and fix the issue for more than just build_runner)?

**Comment by KiddoV** [comment]
Just so you guys know, my `VSCode` was restarted today and the error just disappeared, so I don't know.

**Comment by DanTup** [comment]
@KiddoV does it come back if you re-run build_runner?

**Comment by davidmorgan** [comment]
Note that build_runner actively checks existing content before writing; so if there is no change, it will do no write.

You could delete the output file then build to force a new build.

**Comment by davidmorgan** [comment]
> Yeah, that's what I wondered [above](https://github.com/dart-lang/sdk/issues/28895#issuecomment-5606245567). In which case, maybe a fix in `watcher` to generate events with the natural case of the path would also help (and fix the issue for more than just build_runner)?

Maybe. It's nontrivial; the events that `package:watcher` gets from the VM use the requested case, not the natural case, so either we need to build a new VM route to getting the natural path or `package:watcher` needs to do additional work to discover and track the natural case. I could probably throw something together to check if it fixes the issue once we have a specific repro.

**Comment by KiddoV** [comment]
@DanTup 
That would be annoying if it comes back, but I ran it anyway...
I have tried:
```cmd
dart run build_runner build --delete-conflicting-outputs
```
And also...
```cmd
dart run build_runner clean
dart run build_runner build
```
After that I also "Restart Dart Analysis Server", and "Reload Window" everything comes back perfectly **without** the error.

**Comment by davidmorgan** [comment]
> [@DanTup](https://github.com/DanTup) That would be annoying if it comes back, but I ran it anyway... I have tried:
> 
> dart run build_runner build --delete-conflicting-outputs
> And also...
> 
> dart run build_runner clean
> dart run build_runner build
> After that I also "Restart Dart Analysis Server", and "Reload Window" everything comes back perfectly **without** the error.

None of those should make it come back; even from a clean build build_runner will notice that the outputs haven't changed.

You would need to delete the output files, e.g. `*.g.dart` if you're running a builder that outputs `.g.dart` files, then run build_runner. Then maybe it breaks :)

Or, change an input file so something new gets generated.
